### PR TITLE
Introduce docs on adding new facades

### PR DESCRIPTION
**Pull request by paulirish**
to clear on authorship: @adamraine wrote most of this. I revised it and yote it over here.

ref https://github.com/GoogleChrome/lighthouse/pull/11290

### CODE REVIEW

**Comment by paulirish** [comment]
@addyosmani .. me patrick and adam revised this a bit and i brought it over here. holla if you have suggestions.

File: facades.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    * For example: a YouTube embed iframe.
+* **facade**: a frontend component/library/project of HTML/CSS/JS/etc that 
+  mimics the look of a third-party product and lazily loads it upon user 
+  interaction. It's expected to be significantly more lightweight/faster.
+    * For example: a [lite-youtube-embed](https://github.com/paulirish/lite-youtube-embed).
+
+## Criteria
+### Functions as a facade
+
+* Loads a HTML component on page load which looks like the actual third-party
```
Comment (patrickhulce) [inline facades.md:20]:
```suggestion
* Loads an HTML component on page load which looks like the actual third-party 
```

File: facades.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+### Small Payload
+
+* Should be significantly smaller in comparison to the deferred third-party 
+  product.
+    * The JS/CSS/image/etc payload should be the minimum size necessary to mimic 
+      the product. (A rough rule of thumb is a facade should be &lt; 10% of the 
+      product it's mimicking)
+    * We'd expect the facade to score 99+ in Lighthouse if it's tested on an 
+      otherwise-blank page.
+    * Any required images should be small as is reasonable. Using blurred, low-res,
```
Comment (patrickhulce) [inline facades.md:47]:
```suggestion
    * Any required images should be as small as is reasonable. Using blurred, low-res,
```

File: facades.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# Adding a new facade to third-party-web
+
```
Comment (patrickhulce) [inline facades.md:2]:
```suggestion

In addition to identifying entities and products, third-party-web also includes data on available alternative libraries for products known as _facades_. A _facade_ is a static element which looks similar to the actual embedded resource, but is not functional and therefore much less taxing on the page load.

```

File: facades.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+Lighthouse 7.0 includes a new audit to flag third-party
```
Comment (patrickhulce) [inline facades.md:3]: this paragraph feels like it might belong in lighthouse or web.dev pointing to this doc?

here it feels like just a sentence saying this powers lighthouse would work

**Review by patrickhulce — COMMENTED** [review]
meat of this looks fantastic 👌

File: facades.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+Lighthouse 7.0 includes a new audit to flag third-party 
+embeds (link coming soon) that can be lazy loaded with a facade. The 
+documentation lists recommended facades. If you'd like to add a facade to the 
+list of recommendations, please read the guidance below.
+
+## Terminology
+
+* **third-party product:** a third-party resource that is included on a page.
+    * For example: a YouTube embed iframe.
+* **facade**: a frontend component/library/project of HTML/CSS/JS/etc that
```
Comment (adamraine) [inline facades.md:12]: I think we can shorten this to just "a frontend component of...". A facade bing a library or project is pretty much implied.

Comment (patrickhulce) [reply facades.md:12]:
```suggestion
* **facade**: a frontend component of HTML/CSS/JS/etc that 
```

File: facades.md (hunk 3 above)
Comment (paulirish) [reply facades.md:2]: sure!

File: facades.md (hunk 4 above)
Comment (paulirish) [reply facades.md]: sure. i tweaked it.

since we're spread across LH report, web.dev docs, this doc, etc.. I still want a little bit of context everywhere. it's redundant but whatever.

File: facades.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
+# Facade data in third-party-web
+
+In addition to identifying entities and products, third-party-web also includes data on available alternative libraries for products known as _facades_. A _facade_ is a static element which looks similar to the actual embedded resource, but is not functional and therefore much less taxing on the page load.
+
+[Lighthouse](https://github.com/GoogleChrome/lighthouse/find/master) 7.0 has a new audit that identifies third-party
```
Comment (patrickhulce) [inline facades.md:5]:
```suggestion
[Lighthouse](https://github.com/GoogleChrome/lighthouse) 7.0 has a new audit that identifies third-party 
```

was the `find/master` intentional? seemed like an accident

**Review by patrickhulce — APPROVED** [review]

File: facades.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
+# Facade data in third-party-web
+
+In addition to identifying entities and products, third-party-web also includes data on available alternative libraries for products known as _facades_. A _facade_ is a static element which looks similar to the actual embedded resource, but is not functional and therefore much less taxing on the page load.
```
Comment (adamraine) [inline facades.md:3]: Defining "facade" here seems redundant since we have the terminology section below.

File: facades.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  interaction. It's expected to be significantly more lightweight/faster.
+    * For example: a [lite-youtube-embed](https://github.com/paulirish/lite-youtube-embed).
+
+## Criteria for adding a new facade
+### Basic functionality
+
+* Loads an HTML component on page load which looks like the actual third-party 
+  embed.
+* Replaces that component with the actual third-party embed after some user 
+  interaction
```
Comment (adamraine) [inline facades.md:23]:
```suggestion
  interaction.
```

**Review by adamraine — COMMENTED** [review]
Overall LGTM, thanks for for the help refining this!

File: facades.md (hunk 7 above)
Comment (paulirish) [reply facades.md:3]: agree. @patrickhulce cool if we drop it here?  or we just inline the terminology into sentences up here.

Comment (patrickhulce) [reply facades.md:3]: inline up here SGTM. I wanted to intro to the purpose of this document for the rest of the community and contributors that wasn't "Lighthouse has a new audit" :) 

terminology section was helpful for broad context google doc, but can just be in intro here SGTM

File: facades.md [hunk 9]
```diff
@@ -1,19 +1,10 @@
 # Facade data in third-party-web
 
-In addition to identifying entities and products, third-party-web also includes data on available alternative libraries for products known as _facades_. A _facade_ is a static element which looks similar to the actual embedded resource, but is not functional and therefore much less taxing on the page load.
+In addition to identifying entities, third-party-web also includes data on entities' _products_, and available alternative libraries for products known as _facades_. A **third-party product** is a third-party resource that is included on a page, for example: a YouTube embed iframe. A **facade** is a frontend element which looks similar to the actual product and lazily loads it upon user interaction.  It's expected to be significantly more lightweight/faster. For example: [lite-youtube-embed](https://github.com/paulirish/lite-youtube-embed).
```
Comment (patrickhulce) [inline facades.md:3]:
```suggestion
In addition to identifying entities, third-party-web also includes data on entities' _products_, and available alternative libraries for products known as _facades_. A **third-party product** is a third-party resource that is included on a page associated with a specific integration/SDK/API, for example: a YouTube embed iframe. A **facade** is a frontend element which looks similar to the actual product and lazily loads it upon user interaction.  It's expected to be significantly more lightweight/faster. For example: [lite-youtube-embed](https://github.com/paulirish/lite-youtube-embed).
```

Comment (patrickhulce) [reply facades.md:3]: or something along those lines? otherwise it's not really distinguished within the tpw context from a regular entity resource :)

**Review by patrickhulce — APPROVED** [review]
thanks a lot for working this up @adamraine and @paulirish !

File: facades.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+In addition to identifying entities, third-party-web also includes data on entities' _products_, and available alternative libraries for products known as _facades_. A **third-party product** is a third-party resource that is included on a page, for example: a YouTube embed iframe. A **facade** is a frontend element which looks similar to the actual product and lazily loads it upon user interaction.  It's expected to be significantly more lightweight/faster. For example: [lite-youtube-embed](https://github.com/paulirish/lite-youtube-embed).
+
+[Lighthouse](https://github.com/GoogleChrome/lighthouse) 7.0 has a new audit that identifies third-party 
+embeds (link coming soon) that can be lazy loaded with a facade. The Lighthouse audit is powered by the facade data in third-party-web.
+
+## Criteria for adding a new facade
+### Basic functionality
+
+* Loads an HTML component on page load which looks like the actual third-party
```
Comment (addyosmani) [inline facades.md:11]: Imo, a facade should present an approximate of the actual third-party, but it doesn't strictly need to load an HTML component to do this. It's valid if you have an image/screenshot/generated scrape of the 3P that loads the real thing on interaction.

Something like https://twitter.com/wongmjane/status/1330273157245243394 is an example of where an image can be fine here.

File: facades.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+* Loads an HTML component on page load which looks like the actual third-party 
+  embed.
+* Replaces that component with the actual third-party embed after some user 
+  interaction.
+    * User clicks on it, mouses over, etc.
+
+### Well Maintained
+
+* The projects issues and contributions are managed responsibly. Bugs are
```
Comment (addyosmani) [inline facades.md:19]: Minor: Is there any concern we aren't defining what we mean by `swiftly` here? e.g perhaps we could provide an example (`swiftly (e.g within 30-60 days of filing)`).

File: facades.md [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  embed.
+* Replaces that component with the actual third-party embed after some user 
+  interaction.
+    * User clicks on it, mouses over, etc.
+
+### Well Maintained
+
+* The projects issues and contributions are managed responsibly. Bugs are 
+  handled swiftly.
+* Is already used in production, ideally by users other than the creator.
```
Comment (addyosmani) [inline facades.md:21]: Would this requirement disqualify some of the options we were looking to recommend earlier? e.g Alex's vimeo lite component. I'm not saying this shouldn't be a recommendation (rather, I think production use is a great requirement), just pointing this out... :)

File: facades.md [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      techniques may be advantageous.
+
+## Submission Process
+
+Anyone can propose changes, by making a pull request to this repo. Proposals 
+don't necessarily need to be submitted by the creator.
+
+Read 
+[contributing](https://github.com/patrickhulce/third-party-web#contributing) for 
+basics; look in `data/entities.js` for `products` to see how the data is structured.
```
Comment (addyosmani) [inline facades.md:49]: Rest of this looks great 👍

File: facades.md (hunk 10 above)
Comment (patrickhulce) [reply facades.md:11]: Isn't your example an HTML component?

I interpreted "HTML component" here to mean anything that shows up in the DOM as compared to a library/utility that only exists in JS.

File: facades.md (hunk 11 above)
Comment (patrickhulce) [reply facades.md:19]: I think we were intentionally vague on these points, @paulirish and I had quite the scuffle as to whether it's better to have these discussions on a case-by-case basis or a settled policy ;)

File: facades.md (hunk 12 above)
Comment (patrickhulce) [reply facades.md:21]: I think it would disqualify many, but I think that's why they're guidelines w/case-by-case determination instead of strict requirements and the intention was just to not broadcast that.

File: facades.md (hunk 10 above)
Comment (paulirish) [reply facades.md:11]: yeah i was going with a loose definition of "component". There doesn't feel like there's a great word for this thing. It's not a "library".... even "widget" is close but doesn't feel right.

File: facades.md (hunk 11 above)
Comment (paulirish) [reply facades.md:19]: yeah i just thinking we may be adding more governance than necessary. but `swiftly (e.g within 30-60 days of filing)` wfm.

File: facades.md (hunk 10 above)
Comment (addyosmani) [reply facades.md]: Patrick: after taking a second look, you're of course correct. I know finding the right wording here is tricky and so HTML component may be totally fine if its to be interpreted with a loose definition.

Comment (patrickhulce) [reply facades.md]: If it took this conversation to explain thats what we're talking about to someone already very familiar with the area though, perhaps thats enough of a signal that others might not understand what we're talking about :)

Is there an alternative term you had in mind @addyosmani that has a better blend of the generic yet specific we need here?

**Comment by adamraine** [comment]
Hasn't been movement here in a bit. Any objections to merging?

**Comment by patrickhulce** [comment]
thanks for the reminder @adamraine, done :)
